### DISCUSSION TITLE
Cannot assertSee() changes on nested components after using set()

### DISCUSSION THREAD

**Discussion by legecha** [livewire/livewire#8101] [7. Bugs (archived)]
### Livewire version

v3.4.8

### Laravel version

v10.47

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Create two Livewire components (a parent and child)
2. Create a public property `$name` on both, with `#[Reactive]` on the child
3. Create an input with `wire:model.live="name"` in the parent view
4. Output `$name` in the child view
5. Write a test that sets the name with `->set('name', 'Something')` and then `->assertSee('Something')` and see it fail

### Repository or Wirebox URL

https://github.com/legecha/livewire-testing

### Additional information

**Note:** The repos above has the necessary example code and a test living at `tests/Feature/Livewire/DaddyTest.php` to illustrate the issue.

It appears that in tests the same level of refreshing or morphing doesn't occur. In the example repo provided, typing in the input will update the output in the child component, but in tests it just sees the parent component's HTML with the empty component div, i.e.

```
<div wire:id="Lyp2MFqw5lESKLKieEwu">\n
    <div wire:id="4m8cCWx3t212iNQKnW3P"></div>    <p>\n
        <input type="text" wire:model.live="name" />\n
    </p>\n
</div>\n
```

Note that this isn't just when using reactive props. I also tried it in a loop where the `name` is passed directly into the `view()` call instead - again, works in the browser but not in tests.

At first I thought that testing nested components in this manner at all was impossible, but then forcing a random key (even when not in a loop) each time does actually make the test pass, i.e. in the parent (daddy) component, include the child (son) component like so:

```
<livewire:son :$name key="{{ random_bytes(8) }}" />
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by billyonecan** [comment]
The crux of the issue appears to be that child component placeholders are not _morphed_ when testing. 

The relevant documentation states:

> An important question you might be asking yourself is: what happens when the parent component re-renders and encounters the child show-post components? How will it re-render the children if their snapshots haven't been included in the request payload?
>
>The answer is: they won't be re-rendered.
>
>When Livewire renders the Posts component, it will render placeholders for any child components it encounters.
>
>Here is an example of what the rendered HTML for the Posts component might be after the above update:
>
>```
><div wire:id="123">
>    Post Limit: <input type="number" wire:model.live="postLimit">
> 
>    <div wire:id="456"></div>
></div>
>```
>
>As you can see, only one child has been rendered because postLimit was updated to 1. However, you will also notice that instead of the full child component, there is only an empty `<div></div>` with the matching wire:id attribute.

..and this is as far as it gets (when testing). The following does not occur:

>When this HTML is received on the frontend, Livewire will morph the old HTML for this component into this new HTML, but intelligently skip any child component placeholders.
>
>The effect is that, after _morphing_, the final DOM content of the parent Posts component will be the following:
>
>```
><div wire:id="123">
>    Post Limit: <input type="number" wire:model.live="postLimit">
> 
>    <div wire:id="456">
>        <h1>The first post</h1>
> 
>        <p>Post content</p>
> 
>        <button wire:click="$refresh">Refresh post</button>
>    </div>
></div>
>```

**Reply by joshhanley** [reply]
@billyonecan yeah that is correct, Livewire when it renders a component in a browser, on the second request, none of the child components are rendered, due to them being isolated.

The exact same process happens in a test, so you won't see any UI changes for any of the children, instead all there will be is the placeholder that Livewire uses to denote that a component lives there in the DOM.

**Reply by legecha** [reply]
Presumably this means that, replacing the placeholder with the original (unchanged) content from before the request is the job of the client-side JavaScript, hence why we see the content show up correctly in the browser, but not in the test?

As we can force the children to also be updated, by supplying a new key to the component, would there be a way to achieve this kind of test without having to add them to the blade that will be used in the browser, to try make the behaviour consistent? Or maybe a different method altogether?

Currently, if I extract content away to reusable components and nest them, I can't use the same assertions in my tests to make sure the content is still being displayed properly after the parent content renders again.

**Comment by legecha** [comment]
A potential workaround that won't affect your Blade code too much.

In `app/helpers.php`:

```
if (! function_exists('lkey')) {
    /**
     * Change a given key to a random one when in testing mode.
     */
    function lkey(string $key): string
    {
        return App::environment() === 'testing'
            ? 'test-'.Str::random(8)
            : $key;
    }
}
```

Then in your blade:

```
<livewire:some.component :key="lkey($your->id)" />

```

This doesn't affect it in the browser, but forces a new and unique key when testing, therefore bringing the content back. Tests are now passing.

Still welcome other ideas or knowing if this is something that can be fixed natively.
